goodai/ltm/memory.py

```diff
@@ -0,0 +1,209 @@
+import json
+import time
+import queue
+from typing import Any
+from copy import deepcopy
+from collections import defaultdict
+from goodai.ltm.mem.auto import AutoTextMemory, DefaultTextMemory
+from goodai.ltm.mem.config import TextMemoryConfig
+from goodai.ltm.mem.base import RetrievedMemory, PassageInfo
+from multiprocessing import Queue, Process
+
+
+def memory_server(
+    mem_kwargs: dict, time_budget: float, query_queue: Queue, add_queue: Queue,
+    bkg_queue: Queue, processed_queue: Queue, out_queue: Queue,
+):
+    ltm = LTMSystem(**mem_kwargs)
+    allowed_methods = {"is_empty", "clear", "state_as_text", "set_state", "retrieve",
+                       "retrieve_from_keywords"}
+    while True:
+        t0_loop = time.time()
+        # Attend queries and other priority actions
+        try:
+            query_dict = query_queue.get(block=False)
+            assert query_dict["method"] in allowed_methods
+            return_value = getattr(ltm, query_dict["method"])(**query_dict["kwargs"])
+            out_queue.put(return_value)
+        except queue.Empty:
+            pass
+        # Give additions and changes some slack. Otherwise, queries might hoard time.
+        while True:
+            queues_empty = True
+            t0_changes = time.time()
+            # Add memories and send it to background processing
+            # TODO: process these in random order or randomly drop some, to avoid bottlenecks
+            try:
+                kwargs = add_queue.get(block=False)
+                content, metadata = ltm.content_addition_preprocessing(**kwargs)
+                text_key = ltm.add_content(**kwargs)
+                bkg_kwargs = dict(text=content, metadata=metadata, text_key=text_key)
+                bkg_queue.put(bkg_kwargs)
+                queues_empty = False
+            except queue.Empty:
+                pass
+            # Take processed memories and update memory database
+            try:
+                kwargs = processed_queue.get(block=False)
+                ltm.semantic_memory.replace_text(**kwargs)
+                queues_empty = False
+            except queue.Empty:
+                pass
+            t_changes = time.time() - t0_changes
+            # See if there's time for another round
+            if queues_empty or time.time() - t0_loop + t_changes >= time_budget:
+                break
```

Comment:
The idea is to see how much time we've spent so far, and see if doing another iteration of the _changes_ loop might be within the budget limits. So:
```python
total_time_spent = time.time() - t0_loop
last_change_duration = time.time() - t0_changes  # This is included in total_time_spent
time_after_next_iter = total_time_spent + last_change_duration  # This is an estimation
```

I agree that I could do a better estimation by taking an average or something, but I didn't want to complicate the code even more than it is now. Your suggestion is simpler, but it violates the time constraint. I can implement it though, and just change the terminology to something else.

Reply:
Ahhh, okay I didn't appreciate that. So that's why it only cares about the last push/pop. The `time_budget` constraint could still be violated since we have no guarantee on how long the final push/pop will take, but okay I get it now.

Reply:
yeah, that's a possible violation, specially if the time budget is too low, but I didn't know how to overcome it. The one thing I did was to make sure that in the worst case scenario, it works like a simple round-robin.